subprojects/build-cache/src/main/java/org/gradle/caching/internal/controller/NextGenBuildCacheController.java

```diff
@@ -361,14 +364,31 @@ public void startUnpack(long totalSize) {
                     .details(new UnpackOperationDetails(manifestKey, totalSize))
                     .progressDisplayName("Unpacking build cache entry");
             });
+            if (parentOp == null) {
+                parentOp = unpackBuildOp;
+            }
         }
 
         @Override
         public void close() {
-            loadBuildOp.finish(context -> context.setResult(missEncountered.get()
+            // Load and unpack operations can be started in either order, but then we have to finish them in the same order
+            if (parentOp == loadBuildOp) {
+                finishUnpack();
+                finishLoad();
+            } else {
+                finishLoad();
```

Comment:
Do we even have load operation if unpack is started first? 

If I understand code correctly, now since we do all operation starts in one thread, we have these two options:
1. at least one entry is loaded remotely: we start load operations first and unpack second
2. no entry is loaded remotely: we start only unpack operation, there is no load operations

So: what happens if we call loadFinish(), but there is none load operation started?

Reply:
> Do we even have load operation if unpack is started first?

Yes, this is possible, if the local cache has the manifest, but misses some of the content pieces (presumably because of cleanup).

> So: what happens if we call loadFinish(), but there is none load operation started?

`OnDemandBuildOperationWrapper.finish()` would be called and execute a no-op because the internal `context` is `null`.